nearcore/src/lib.rs

```diff
@@ -74,17 +71,89 @@ pub fn get_default_home() -> PathBuf {
     PathBuf::default()
 }
 
+/// Returns the path of a directory where DB checkpoint can be found or created.
+/// Default location is the same as the database location: `path`.
+fn db_checkpoint_root_dir<'a>(path: &'a Path, near_config: &'a NearConfig) -> &'a Path {
+    if let Some(db_checkpoints_path) = &near_config.config.db_checkpoints_path {
+        &db_checkpoints_path
+    } else {
+        path
+    }
+}
+
+/// Returns the path of the DB checkpoint.
+fn db_checkpoint_path<'a>(path: &'a Path, near_config: &'a NearConfig) -> PathBuf {
+    let checkpoint_path = db_checkpoint_root_dir(path, near_config);
+    let checkpoint_path = checkpoint_path.join(DB_CHECKPOINT_NAME);
+    checkpoint_path
+}
+
+const DB_CHECKPOINT_NAME: &str = "db_snapshot_before_migration";
+
+/// Returns the path of the first DB checkpoint found on disk.
+fn check_db_checkpoint_exists(path: &Path, near_config: &NearConfig) -> Option<PathBuf> {
+    let path = db_checkpoint_path(path, near_config);
+    if path.exists() {
+        Some(path)
+    } else {
+        None
+    }
+}
+
+/// Creates a consistent DB checkpoint and returns its path.
+/// By default it creates checkpoints in the DB directory, but can be overridden by the config.
+fn create_db_checkpoint(path: &Path, near_config: &NearConfig) -> Result<PathBuf, anyhow::Error> {
+    let checkpoint_path = db_checkpoint_path(path, near_config);
+
+    let db = RocksDB::new(path)?;
+    let checkpoint = db.checkpoint()?;
+    info!(target:"near","Creating a DB snapshot in '{}'",checkpoint_path.display());
+    checkpoint.create_checkpoint(&checkpoint_path)?;
+    info!(target:"near","Created a DB snapshot in '{}'",checkpoint_path.display());
+
+    Ok(checkpoint_path)
+}
+
 /// Function checks current version of the database and applies migrations to the database.
 pub fn apply_store_migrations(path: &Path, near_config: &NearConfig) {
     let db_version = get_store_version(path);
     if db_version > near_primitives::version::DB_VERSION {
         error!(target: "near", "DB version {} is created by a newer version of neard, please update neard or delete data", db_version);
         std::process::exit(1);
     }
+
     if db_version == near_primitives::version::DB_VERSION {
         return;
     }
 
+    // Before attemping a DB migration check if the current DB is corrupted. We assume that
+    // checkpoints get created before each migration attempt and get removed after successful
+    // attempts.
+    if near_config.config.use_checkpoints_for_db_migration {
+        match check_db_checkpoint_exists(path, near_config) {
+            Some(existing_checkpoint_path) => {
+                panic!("Detected an existing DB migration checkpoint: '{}'. Probably a DB migration got interrupted and your DB is corrupted. Please replace the contents of '{}' with data from that checkpoint, delete the checkpoint and try again.",existing_checkpoint_path.display(), path.display());
```

Comment:
This is quite hard to read on a finite display, you can wrap this using `\`: 

```
        match check_db_checkpoint_exists(path, near_config) {
            Some(existing_checkpoint_path) => {
                panic!(
                    "Detected an existing DB migration checkpoint: '{}'. \
                     Probably a DB migration got interrupted and your DB is corrupted. \
                     Please replace the contents of '{}' with data from that checkpoint, \
                     delete the checkpoint and try again.",
                    existing_checkpoint_path.display(),
                    path.display()
                );
            }
            None => {}
        }
```

We might also hard wrap this (with `\n`), as terminals are often bad at wrapping, but I don't know what's the best practice here.